core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java

```diff
@@ -103,30 +108,48 @@ public CoreEvent process(CoreEvent event) throws MuleException {
-        final CoreEvent event = me.getEvent();
-
-        if (me.getFailingComponent() != null &&
-            isWithinSourcePolicy(me.getFailingComponent().getLocation())) {
-          return policyEventMapper.fromPolicyNext(event);
-        } else {
-          return policyEventMapper.fromPolicyNext(policyEventMapper.onFlowError(event, getPolicyId(),
-                                                                                SourcePolicyContext.from(event)
-                                                                                    .getParametersTransformer()));
-        }
-      }, notificationHelper, getLocation());
-    } else {
-      return new OnExecuteNextErrorConsumer(me -> policyEventMapper.fromPolicyNext(me.getEvent()), notificationHelper,
-                                            getLocation());
-    }
+    return new OnExecuteNextErrorConsumer(me -> {
+      if (!isWithinSourcePolicy(getLocation(), me)) {
+        return policyEventMapper.fromPolicyNext(me.getEvent());
+      }
+      final CoreEvent event = me.getEvent();
+
+      if (me.getFailingComponent() != null && isWithinSourcePolicy(me.getFailingComponent().getLocation(), me)) {
+        return policyEventMapper.fromPolicyNext(event);
+      } else {
+        return policyEventMapper.fromPolicyNext(policyEventMapper
+            .onFlowError(event, getPolicyId(), SourcePolicyContext.from(event).getParametersTransformer()));
+      }
+
+    }, notificationHelper, getLocation());
   }
 
-  private Boolean isWithinSourcePolicy(final ComponentLocation loc) {
+  private Boolean isWithinSourcePolicy(final ComponentLocation loc, MessagingException me) {
     return loc.getParts().size() >= 2
-        && loc.getParts().get(1).getPartIdentifier()
+        && (loc.getParts().get(1).getPartIdentifier()
             .map(tci -> tci.getIdentifier().getName().equals(SOURCE_POLICY_PART_IDENTIFIER))
-            .orElse(false);
+            .orElse(false) || isWithingSubflowInSourcePolicy(loc, me));
+  }
+
+  private Boolean isWithingSubflowInSourcePolicy(final ComponentLocation loc, MessagingException me) {
+    if (loc.getParts().size() < 1) {
+      return false;
+    }
+    boolean isWithinSubflow = loc.getParts().get(0).getPartIdentifier()
+        .map(tci -> tci.getIdentifier().getName().equals(SUBFLOW_POLICY_PART_IDENTIFIER)).orElse(false);
+    if (!isWithinSubflow) {
+      return false;
+    }
+    List<FlowStackElement> elements = me.getEvent().getFlowCallStack().getElements();
+    if (elements.size() != 1) {
+      return false;
+    }
+
+    Location completeLocation = builderFromStringRepresentation(elements.get(0).getProcessorPath().split(" ")[0]).build();
+    List<String> parts = completeLocation.getParts();
+    boolean isSubflowWithinASource = completeLocation.getGlobalName().equals(getRootContainerLocation().toString())
+        && parts.size() >= 1 && parts.get(0).equals(SOURCE_POLICY_PART_IDENTIFIER);
+    return isSubflowWithinASource;
```

Comment:
return directly, without aux var

Reply:
I added the variable just for legibility. I can remove it, tho